provisioner/terraform/tfparse/tfextract.go

```diff
@@ -1,40 +1,64 @@
 package tfparse
 
 import (
+	"archive/zip"
+	"bytes"
 	"context"
 	"encoding/json"
+	"io"
 	"os"
 	"slices"
 	"sort"
+	"strconv"
 	"strings"
 
+	"github.com/coder/coder/v2/archive"
+	"github.com/coder/coder/v2/provisionersdk"
 	"github.com/coder/coder/v2/provisionersdk/proto"
 
 	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/hcl/v2/hclparse"
 	"github.com/hashicorp/hcl/v2/hclsyntax"
 	"github.com/hashicorp/terraform-config-inspect/tfconfig"
+	"github.com/zclconf/go-cty/cty"
+	"golang.org/x/exp/maps"
 	"golang.org/x/xerrors"
 
 	"cdr.dev/slog"
 )
 
-// WorkspaceTags extracts tags from coder_workspace_tags data sources defined in module.
-func WorkspaceTags(ctx context.Context, logger slog.Logger, module *tfconfig.Module) (map[string]string, error) {
-	workspaceTags := map[string]string{}
+// NOTE: This is duplicated from coderd but we can't import it here without
+// introducing a circular dependency
+const maxFileSizeBytes = 10 * (10 << 20) // 10 MB
+
+// ParseHCLFile() is only method of hclparse.Parser we use in this package.
+// hclparse.Parser will by default cache all previous files it has ever
+// parsed. While leveraging this caching behavior is nice, we _never_ want
+// to end up in a situation where we end up returning stale values.
+type Parser interface {
+	ParseHCLFile(filename string) (*hcl.File, hcl.Diagnostics)
+}
 
+// WorkspaceTags extracts tags from coder_workspace_tags data sources defined in module.
+// Note that this only returns the lexical values of the data source, and does not
+// evaluate variables and such. To do this, see evalProvisionerTags below.
+// If the provided Parser is nil, a new instance of hclparse.Parser will be used instead.
+func WorkspaceTags(ctx context.Context, logger slog.Logger, parser Parser, module *tfconfig.Module) (map[string]string, error) {
```

Comment:
This might be the right time to perform a small refactor and convert all these functions in a proper `TerraformParser`. 

```go
parser := NewTerraformParser(logger) // wil set an instance of HCLParser
... := parser.ParseWorkspaceTags(ctx, module)

... := parser.ParseWorkspaceTagDefaultsFromFile(ctx,  file, mimetype) // can reuse instance of HCLParser
```